statsmodels/tsa/vector_ar/var_model.py

```diff
@@ -926,22 +925,33 @@ class VARProcess(object):
         are the variable and the rows are the equations so that coefs[i-1] is
         the estimated A_i matrix. See Notes.
     coefs_exog : ndarray
-        parameters for trend and user provided exog
+        1d or 2d array. If 1d, should be of length neqs and is assumed to be
+        a vector of constants. If 2d should be of shape k_trend x neqs
     cov_resid : ndarray (neqs x neqs)
         The covariance matrix of the residuals. :math:`\Sigma_u` in the Notes.
-    names : sequence (length k)
-    _params_info : dict
-        internal dict to provide information about the composition of `params`,
-        specifically `k_trend` (trend order) and `k_exog_user` (the number of
-        exog variables provided by the user).
-        If it is None, then coefs_exog are assumed to be for the intercept and
-        trend.
+    names : sequence (length neqs)
+        The names of the endogenous variables.
+    k_trend : int, optional
+        Order of the trend
+    k_exog_user : int, optional
+        Number of exogenous variables excluding the trend in the model
 
-    Returns
-    -------
-    **Attributes**:
+    Notes
+    -----
+    The VAR(p) process is assumed to be
+
+    .. math::
+
+       y_t = A_1 y_{t-1} + \\ldots + A_p y_{t-p} + u_t
+
+    where
+
+    .. math::
+
+       u_t \sim MVN(0, \Sigma_u)
     """
-    def __init__(self, coefs, coefs_exog, cov_resid, names=None, _params_info=None):
+    def __init__(self, coefs, coefs_exog, cov_resid, names=None,
+                 k_trend=None, k_exog_user=None):
```

Comment:
I had introduced `_params_info` to move away from having to keep the info about all individual pieces of params around. For example some models switched to using a Parameter class to have params handling in a central object.
VECM has also deterministic seasonal dummies which are not (yet) in VAR.

During VECM merge, only the basic exog support was added to VAR. The full exog and deterministic support in VAR still needs work.  One possibility is to merge trend, seasonal dummies into exog, but in VECM that has the problem that inference and stochastic properties are not the same for deterministic and weakly exogenous variables.

Reply:
The real problem is that the hierarchy between VARProcess and VAR is wrong.  I don't think this is the PR to fix that issue.  It is better to get it cleaned up a lot and bring in some of the existing PRs rather than to get bogged down trying to do a big refactor at the same time.

Reply:
But whatever refactoring is still coming, IMO, the `_params_info` is more flexible and more future proof/resistant, than switching back to using individual k_xxx in the signature.

e.g. where do we put deterministic seasonal of VECM?
My guess is that we need or can use an `exog_all` that combines, trend/constant, exog_user and deterministic, for those places when they can all be treated in the same way.
However, for forecasting we can build trend and seasonal components directly, but users need to provide exog_user.

Reply:
VARProcess should know about trends that can be well defined, e.g., 'c' or 'ct', but general trend arrays should not be passed to a VARProcess since these aren't consistent with defining a 'process'.  A process should, for example, be simulable from its definition. There probably needs to be an intermediate, hidden class, after VARProcess and before other user-visible classes that add more general support of VAR-X type specifications.

Reply:
I don't see anything special about 'c' or 'ct' trend. Anything that changes the unconditional mean in a period can be lumped together for simulation.
e.g. we can have a predict_mean method that creates the exog/deterministic part which can include seasonal components, known structural breaks in trend, future exog that are policy variables and set outside of the model.
I don't see why the "process" would care about what drives the "strictly exogenous" mean shift in each period.

Reply:
'nc','c', and 'ct' are special in the sense that they, in combination with model parameters and a distribution of the shocks, provide a complete description of a stochastic process.  A (nobs, q) array of exog "data" without the structure of a complete process behind it, do not.   IOW passing 'c' is providing an exog "generator' that can generate draws from the exogenous process for any number of time periods or across many simulations. These definitions allow the question as to whether the process is_stable to be correctly answered since the process is well defined. 

> I don't see why the "process" would care about what drives the "strictly exogenous" mean shift in each period.

This isn't a complete description of a stochastic process without another "process" the describes the exogenous shifts.

Reply:
> But whatever refactoring is still coming, IMO, the _params_info is more flexible and more future proof/resistant, than switching back to using individual k_xxx in the signature.

I agree with @bashtage on this one (recall I objected to this pattern back when it was first implemented)

`_params_info` is functionally equivalent to `**kwargs`, with the main difference that it is non-standard so introduces cognitive overhead for non-Josef readers (and makes future maintenance more Josef-constrained).  It also does no validation on passed keys, a point on which the explicit version in this PR is strictly superior.

If this is really future-proof/resistant, why is this pattern not used elsewhere?  I don't like the `**kwargs` pattern either, but at least it is something readers are used to.

> e.g. where do we put deterministic seasonal of VECM?

IIRC from when this was merged, the _only_ extant tests are for deterministic seasonal cases.  These could be captured in a "trend" variable like "s4" or "s12".

Splitting `VARProcess` into `VARProcess` and `VARXProcess(VARProcess)` is not a topic for this PR, but is one that I would definitely support (thematically in line with #5206).  It is much simpler to get the "A process should, for example, be simulable from its definition." behavior without the special-casing within `VARProcess.__init__`

Reply:
In my view moving away from `_params_info` is a regression.
The main disadvantage of kwargs is that they are all main keywords instead of using separate dicts or instances of some classes by functional separation, e.g. I'm using cov_kwds and optim_kwds in some places.

AFAIK, nc, c, ct are only of theoretical interest because it is more easy to get some asymptotics.
seasonal dummies or seasonal splines are theoretically similar and treated as stationary deterministics in VECM.
In practice we also want to allow structural breaks in trend or non-power polynomials like chebycheff.
And we might want to do conditional simulation that takes a specific path of the `exog` as given.
E.g. when we want to simulate temperature specific endog, then conditioning on the termperature path of the last century might be a useful simulation.

At the end of VECM merge, there was still the open question how to define some of the time series properties in the presence of trend or exog, e.g. long-run mean doesn't make sense with trend or trending exog, and I think those process properties including impulse response functions should be defined ignoring all trend, deterministic and exog variables, but those are not ignored for long-term prediction at a finite forecast horizon.